Updating CONTRIBUTING.md for eclipse/che#21363

SDawley opened the pull request:
<!-- Please review the following before submitting a PR:
chectl's Contributing Guide: https://github.com/che-incubator/chectl/blob/main/CONTRIBUTING.md
Pull Request Policy: https://github.com/che-incubator/chectl/blob/main/CONTRIBUTING.md#push-changes-provide-pull-request
-->

### What does this PR do?
Removed Generate Devfile and Create Workspace sections and edited the command documentation to reflect how commands are currently run.

Signed-off-by: Samantha Dawley <sdawley@redhat.com>

### Screenshot/screencast of this PR
<!-- Please include a screenshot or a screencast explaining what is doing this PR -->


### What issues does this PR fix or reference?
<!-- Please include any related issue from eclipse che repository (or from another issue tracker).
     Include link to other pull requests like documentation PR from [the docs repo](https://github.com/eclipse/che-docs)
-->
Addresses: https://github.com/eclipse/che/issues/21363 

### How to test this PR?
<!-- Please explain for example :
  - The test platform (openshift, kubernetes, minikube, CodeReady Container, docker-desktop, etc)
  - steps to reproduce
 -->


### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

tolusha commented:
Sounds good. What about putting `https://console-openshift-console.apps.cluster-n56gt.n56gt.sandbox689.opentlc.com/` url instead of `<your-che-host>` ?

SDawley commented:
> Sounds good. What about putting sandbox url instead of `<your-che-host>` to have defined url to start a workspace ?

OK, switched to the OpenShift Sandbox url.

max-cx commented on CONTRIBUTING.md:
```diff
@@ -3,53 +3,45 @@
+## Start Workspace
 
-To create a workpace you can use [devfile](devfile.yaml):
+To start a workpace from the [devfile](devfile.yaml) create a workspace URL by appending the chectl repo to the URL of your Che host. For example, to create a workspace in the openshift sandbox use the URL:
```
```suggestion
To start a workpace from the [devfile](devfile.yaml), create a workspace URL by appending the chectl repository to the URL of your Che host. For example, to create a workspace in the OpenShift sandbox, use the following URL:
```

benoitf replied:
FYI there is a typo on workpace --> workSpace
```suggestion
To start a workspace from the [devfile](devfile.yaml) create a workspace URL by appending the chectl repo to the URL of your Che host. For example, to create a workspace in the openshift sandbox use the URL:
```

nickboldt replied:
Probably want to capitalize `Sandbox` as seen on https://developers.redhat.com/developer-sandbox

max-cx commented on CONTRIBUTING.md:
```diff
@@ -3,53 +3,45 @@
+> See more about [devfile](https://devfile.io/docs/2.1.0/what-is-a-devfile)
+> See more about [workspace URLs](https://www.eclipse.org/che/docs/stable/end-user-guide/starting-a-new-workspace-with-a-clone-of-a-git-repository/)
 
 After starting the workspace Theia will clone sources of `chectl` to `/projects/chectl` directory.
```
```suggestion
After starting the workspace, Theia will clone sources of `chectl` to the `/projects/chectl` directory.
```

nickboldt replied:
Since we will soon be switching IDE default from Theia to VSCode, perhaps "After starting the workspace, the IDE will clone sources..." is better?

max-cx commented on CONTRIBUTING.md:
```diff
@@ -3,53 +3,45 @@
 
 After starting the workspace Theia will clone sources of `chectl` to `/projects/chectl` directory.
 
 `chectl` is written in TypeScript. For its development there is a dedicated `dev` container which has preinstalled software for comfortable development. That allows the developer to build, test and launch `chectl` inside the container.
```
```suggestion
`chectl` is written in TypeScript. For its development, there is a dedicated `dev` container that has preinstalled software for comfortable development. This allows the developer to build, test, and launch `chectl` inside the container.
```

max-cx commented on CONTRIBUTING.md:
```diff
@@ -3,53 +3,45 @@
-2. `Terminal => Run Task...` menu. You can fin and pick a command in the appeared command palette.
-
-3. Manually by opening a terminal in the `dev` container and running commands in `/projects/chectl` directory.
+You can run commands through the Terminal menu by clicking `Terminal => Run Task...` then selecting the desired task. 
```
```suggestion
You can run thee commands through the **Terminal** menu by clicking **Terminal > Run Task** and selecting the desired task. 
```

max-cx commented on CONTRIBUTING.md:
````diff
@@ -3,53 +3,45 @@
 ```
 
-Open `My Workspace` panel at the left and launch `Build` command. It will run `yarn` command in `/projects/chectl` directory inside `dev` container. The command will install all necessary dependencies and compile the project. Upon successfull assembly, a new `bin` directory will appear in project directory and will have two files: `run` and `run.cmd`.
+Running the  `che: build` command will run `yarn` in the `/projects/chectl` directory inside `dev` container. The command will install all necessary dependencies and compile the project. Upon successfull assembly, a new `bin` directory will appear in project directory and will have two files: `run` and `run.cmd`.
````
```suggestion
Running the  `che: build` command will run `yarn` in the `/projects/chectl` directory inside `dev` container. The command will install all necessary dependencies and compile the project. Upon successful assembly, a new `bin` directory will appear in the project directory and will have two files: `run` and `run.cmd`.
```

max-cx commented on CONTRIBUTING.md:
````diff
@@ -59,7 +51,7 @@ yarn test
 ```
 
 Tests for `chectl` are written using `jest` framework.
````
```suggestion
Tests for `chectl` are written by using the [jest](https://jestjs.io/docs/getting-started) framework.
```

max-cx commented on CONTRIBUTING.md:
```diff
@@ -69,41 +61,11 @@ Testing summary will be printed to the output.
 ## Package binaries
-For packaging binaries, [oclif](https://github.com/oclif/dev-cli) is used. It generates packages for Linux, Windows and MacOS operation systems and puts the result in `dist/channels/stable` directory.
-To start packaging just run `Package Binaries` commands from `My Workspace`. It will run the following in `/projects/chectl` directory.
+For packaging binaries, [oclif](https://github.com/oclif/dev-cli) is used. It generates packages for Linux, Windows and MacOS operation systems and puts the result in the `dist/channels/stable` directory.
```
```suggestion
For packaging binaries, [oclif](https://github.com/oclif/dev-cli) is used. It generates packages for Linux, Windows, and MacOS operation systems and puts the result in the `dist/channels/stable` directory.
```

nickboldt replied:
"operation systems" => "operating systems"

max-cx commented on CONTRIBUTING.md:
```diff
@@ -69,41 +61,11 @@ Testing summary will be printed to the output.
-For packaging binaries, [oclif](https://github.com/oclif/dev-cli) is used. It generates packages for Linux, Windows and MacOS operation systems and puts the result in `dist/channels/stable` directory.
-To start packaging just run `Package Binaries` commands from `My Workspace`. It will run the following in `/projects/chectl` directory.
+For packaging binaries, [oclif](https://github.com/oclif/dev-cli) is used. It generates packages for Linux, Windows and MacOS operation systems and puts the result in the `dist/channels/stable` directory.
+To start packaging just run the `che: package-binaries` command. It will run the following in `/projects/chectl` directory:
```
```suggestion
To start packaging, just run the `che: package-binaries` command. It will run the following in `/projects/chectl` directory:
```

max-cx reviewed (commented):
Just a quick review.

SDawley commented:
Updated including Max, Florent, and Nick's suggested fixes.

tolusha commented:
1. Remove `Create Workspace` and `Generate Devfile` task from the Devfile.
2. I created a workspace on dogfooding instance `https://che-dogfooding.apps.che-dev.x6e0.p1.openshiftapps.com/f?url=https://github.com/che-incubator/chectl` with che-code as a default editor. After running `che: build` task I got:

```
yarn install v1.22.17
[1/5] Validating package.json...
error chectl@0.0.2: The engine "node" is incompatible with this module. Expected version ">=16.0.0". Got "14.19.0"
error Found incompatible module.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

nickboldt commented:
> 1. Remove `Create Workspace` and `Generate Devfile` task from the Devfile.
> 2. I created a workspace on dogfooding instance `https://che-dogfooding.apps.che-dev.x6e0.p1.openshiftapps.com/f?url=https://github.com/che-incubator/chectl` with che-code as a default editor. After running `che: build` task I got:
> 
> ```
> yarn install v1.22.17
> [1/5] Validating package.json...
> error chectl@0.0.2: The engine "node" is incompatible with this module. Expected version ">=16.0.0". Got "14.19.0"
> error Found incompatible module.
> info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
> ```

Strange. The Che UDI image has 3 different vesions of node (12, 14, and 16) so you should be able to build it there using that image:

https://github.com/devfile/developer-images/blob/main/universal/ubi8/Dockerfile#L62-L71

SDawley commented:
I created a workspace and manually updated the image to 'quay.io/devfile/universal-developer-image:ubi8-latest'

The build command now works, as does run but package-binaries is failing so I am investigating.

tolusha commented:
@SDawley 
Pls, convert devfile from `apiVersion: 1.0.0` to `apiVersion: 2.1.0`.

nickboldt reviewed (approved):
Tests are passing so if @tolusha is happy, then +1 to merge.

SDawley commented:
> @SDawley Pls, convert devfile from `apiVersion: 1.0.0` to `apiVersion: 2.1.0`.

@tolusha 
I've updated the devfile to 2.1.0 and added an extensions.json with redhat.vscode-yaml, I did not find che-incubator.typescript in [open-vsx.org](https://open-vsx.org/?search=typescript%20&category=&sortBy=relevance&sortOrder=desc).

I'm able to create a factory link from this branch, but package-binaries still fails with: 
`/bin/sh: shasum: command not found
    Error: Command failed: /bin/sh -c grep node-v16.14.0-linux-x64.tar.xz /projects/chectl/tmp/cache/16.14.0/SHASUMS256.txt.asc | shasum -a 256 -c -
    Code: 127
error Command failed with exit code 1.`
I'm fairly certain its oclif that's requiring shasum. So we can either find a new packaging method or add a symlink to the container from sha265sum to shasum.

Everything else seems to be working, though the tests take a lot longer to run in the workspace than they do locally

tolusha commented:
Have a look
https://open-vsx.org/extension/vscode/typescript-language-features
https://open-vsx.org/extension/vscode/typescript

tolusha commented:
Remove `projects:` from the devfile, otherwise it always clone the main branch from the che-incubator/chectl` repository

tolusha commented:
Have a look at https://github.com/che-samples/react-web-app/blob/devfilev2/devfile.yaml
Pls add `label` and `group`  to the commands.

tolusha commented:
Use `yarn lint:fix` as format source command

tolusha commented on .vscode/launch.json:
```diff
@@ -65,6 +65,14 @@
             "program": "${workspaceFolder}/prepare-che-operator-templates.js"
           },
-
+        {
```
It turns out that we already have `Attach to Remote` configuration in the `launch.json`
So, this changes are unnecessary, pls revert them.
Also, pls remove the following configuration from the `launch.json` as well
* `Debug prepare-che-operator.js script`
* `Debug Openshift Install`

tolusha commented on devfile.yaml:
```diff
@@ -10,113 +10,56 @@
+        kind: build
+  - id: format-sources
+    exec:
+      label: "Run 'yarn format'"
```
```suggestion
      label: "Run 'yarn lint:fix'"
```

tolusha commented on package.json:
```diff
@@ -37,6 +37,7 @@
     "os-locale": "^5.0.0",
     "rimraf": "^3.0.2",
     "semver": "^7.3.7",
+    "shasum": "^1.0.2",
```
```suggestion
```

tolusha commented on package.json:
```diff
@@ -66,8 +67,8 @@
-    "jest": "^26.6.3",
-    "ts-jest": "^26.5.6",
+    "jest": "^29.0.3",
+    "ts-jest": "^29.0.1",
```
Why do we need this?

SDawley replied:
The jest tests were failing in the workspace at version 26 and updating to 29 fixed it. Enough else has changed that I can try reverting that and see if the issue comes back.

tolusha replied:
I tried and did not face any issue with 26.x versions

tolusha commented:
>  So we can either find a new packaging method or add a symlink to the container from sha265sum to shasum.

We definitely need to add `shasum` into a dev container
I spotted another problem: `yarn test` takes forever. I used dogfooding cluster. Did you have the similar issue with `yarn test` ?

SDawley commented:
> > So we can either find a new packaging method or add a symlink to the container from sha265sum to shasum.
> 
> We definitely need to add `shasum` into a dev container I spotted another problem: `yarn test` takes forever. To fix this issue, pls add for a dev container:
> 
> ```
>       cpuLimit: '2'
>       cpuRequest: 500m
>       memoryLimit: 5G
>       memoryRequest: 1G
> ```

I tried adding this to the devfile and restarting the workspace in RHOSDS and got the error:
`Detected unrecoverable deployment condition: FailedCreate pods "workspace8fb0fe516e274bf6-6fbd749ff6-89695" is forbidden: exceeded quota: for-sdawley-compute, requested: limits.memory=6042653184, used: limits.memory=6042653184, limited: limits.memory=7Gi`

tolusha commented:
> I tried adding this to the devfile and restarting the workspace in RHOSDS and got the error:

There is no much we can do here. If it requires at least 5G memory, then we have to use another cluster.
I personally use https://che-dogfooding.apps.che-dev.x6e0.p1.openshiftapps.com/dashboard/

l0rd commented on CONTRIBUTING.md:
```diff
@@ -69,41 +61,11 @@ Testing summary will be printed to the output.
 ## Package binaries
-For packaging binaries, [oclif](https://github.com/oclif/dev-cli) is used. It generates packages for Linux, Windows and MacOS operation systems and puts the result in `dist/channels/stable` directory.
-To start packaging just run `Package Binaries` commands from `My Workspace`. It will run the following in `/projects/chectl` directory.
+For packaging binaries, [oclif](https://github.com/oclif/dev-cli) is used. It generates packages for Linux, Windows, and MacOS operation systems and puts the result in the `dist/channels/stable` directory.
```
```suggestion
For packaging binaries, [oclif](https://github.com/oclif/dev-cli) is used. It generates packages for Linux, Windows, and MacOS operating systems and puts the result in the `dist/channels/stable` directory.
```

dkwon17 commented on CONTRIBUTING.md:
```diff
@@ -69,41 +61,11 @@ Testing summary will be printed to the output.
-For packaging binaries, [oclif](https://github.com/oclif/dev-cli) is used. It generates packages for Linux, Windows and MacOS operation systems and puts the result in `dist/channels/stable` directory.
-To start packaging just run `Package Binaries` commands from `My Workspace`. It will run the following in `/projects/chectl` directory.
+For packaging binaries, [oclif](https://github.com/oclif/dev-cli) is used. It generates packages for Linux, Windows, and MacOS operation systems and puts the result in the `dist/channels/stable` directory.
+To start packaging, just run the `che: package-binaries` command. It will run the following in `/projects/chectl` directory:
```
```suggestion
To start packaging, just run the `che: package-binaries` command. It will run the following in the `/projects/chectl` directory:
```

dkwon17 commented on CONTRIBUTING.md:
```diff
@@ -113,26 +75,26 @@ yarn oclif-dev pack
-There is a required Pull Request check named `Semantic Pull Request` that is ensuring that all commits messages are correctly setup. In order to merge a Pull Request, it has to be green.
+`chectl` uses several Pull Request checks:
+ - The [Conventional Commits](https://conventionalcommits.org) convention for the commit messages.
+There is a required pull request check named **Semantic Pull Request** that ensures that all commits messages are correctly set up. In order to merge a pull request, it has to be green.
```
```suggestion
There is a required pull request check named **Semantic Pull Request** that ensures that all commit messages are correctly set up. In order to merge a pull request, it has to be green.
```

dkwon17 commented on CONTRIBUTING.md:
```diff
@@ -58,8 +50,8 @@ Open `My Workspace` panel at the left and launch `Build` command. It will run `y
-Tests for `chectl` are written using `jest` framework.
-To run tests, go to `My Workspace` panel, find and launch `Test` command. The command will run `yarn test` inside `dev` container.
+Tests for `chectl` are written by using the [jest](https://jestjs.io/docs/getting-started) framework.
+To run tests, find and launch `che: test` command. The command will run `yarn test` inside `dev` container.
```
```suggestion
To run tests, find and launch the `che: test` command. The command will run `yarn test` inside `dev` container.
```

dkwon17 commented on CONTRIBUTING.md:
```diff
@@ -147,4 +109,4 @@ To run the script with docker env
 
-Commit the changes of the script and then, push release branch by overriding current remote release branch
\ No newline at end of file
+Commit the changes of the script and then push to the release branch by overriding current remote release branch.
```
```suggestion
Commit the changes of the script and then push to the release branch by overriding the current remote release branch.
```

dkwon17 reviewed (commented):
Looks great, I was able to build, test and run `chectl` in my workspace. I just have some small suggestions

SDawley commented:
I had to rebuild the universal image because it didn't use the new base image, but packaging binaries now works.
@tolusha

tolusha commented:
I have another error now
```
oclif-dev: building target chectl-win32-x64.tar.gz
/bin/sh: 7z: command not found
 ›   Error: install 7-zip to package windows tarball
error Command failed with exit code 2.
```

SDawley commented:
@tolusha 
I've moved the commands from the devfile to the .vscode/tasks.json since that makes finding and executing them easier. 
Package binaries is working now that 7zip is in ubi8.

tolusha commented:
Sounds good.!

tolusha commented:
Pls, fix DCO issue.
